api/v1beta2/mysqlcluster_types.go

```diff
@@ -259,17 +266,104 @@ func (s MySQLClusterSpec) validateUpdate(old MySQLClusterSpec) field.ErrorList {
 			newSize := pvc.StorageSize()
 			oldSize := old.StorageSize()
 
-			if newSize.Cmp(oldSize) == -1 {
+			switch cmp := newSize.Cmp(oldSize); {
+			case cmp == -1:
 				p := p.Child("volumeClaimTemplates").Index(i).
 					Child("spec").Child("resources").Child("requests").Key("storage")
 				allErrs = append(allErrs, field.Forbidden(p, "storage size cannot be reduced"))
+			case cmp == 1:
+				isVolumeExpansion = true
+			case cmp == 0:
+				// noop
 			}
 		}
 	}
 
+	if isVolumeExpansion {
+		allErrs = append(allErrs, s.validateVolumeExpansionSupported(ctx, apiReader)...)
```

Comment:
I want to stop using the `apiReader` global variable.
Could you create structures for validation?
E.g. `mySQLClusterValidator` and `backupPolicyValidator`.

Recently, we created the following webhook.
I want it to be encapsulated like this.

https://github.com/cybozu-go/nyamber/blob/db1370fcb6443fdfb0f95f7061749db041a35cf2/hooks/virtualdc.go#L34-L45

Reply:
Thanks, I fixed: https://github.com/cybozu-go/moco/pull/420/commits/1912d5b491e93185b18fabd2cc80435cc757e334

I have left the validation logic itself as a method of the MySQLCluster struct, but do you think I should move it to the custom validator side?

Reply:
Sorry for the late reply. I think it is fine as is. Thank you for the good update.